### PR TITLE
Add an LSP server command

### PR DESCRIPTION
**Pull request by mcy**
This PR adds a new command, buf beta lsp, which serves an LSP for Protobuf files.

Most of the code lives under `buflsp`. Currently, the LSP implements:

- Parse and diagnose.
- Format on save.
- Go to definition, inc. go to definition of Buf dependencies.
- Semantic highlighting.
- Symbol tooltips.

### CODE REVIEW

File: private/buf/buflsp/file_manager.go
Comment (bufdev) [inline private/buf/buflsp/file_manager.go:34]: `newFileManager`

File: private/buf/buflsp/file.go
Comment (doriable) [inline private/buf/buflsp/file.go:66]: Is this already addressed?

File: private/buf/buflsp/buflsp.go
Comment (emcfarlane) [inline private/buf/buflsp/buflsp.go:196]: This looks like a data race on `retErr`. I'm not sure you need the go routine here as passing down the context should be enough to ensure the work is bounded.

File: private/buf/buflsp/buflsp.go
Comment (mcy) [reply private/buf/buflsp/buflsp.go:40]: This shouldn't be made configurable until it becomes a problem. It's basically a relief-valve for potential latency issues.

File: private/buf/buflsp/buflsp.go
Comment (mcy) [reply private/buf/buflsp/buflsp.go:196]: No, this is necessary because jsonrpc2 does not parallelize requests by default.

Comment (emcfarlane) [reply private/buf/buflsp/buflsp.go:207]: This go routine doesn't parallelize the processing of requests as it blocks and waits for the completion. The same blocking can be done without the go routine and rely on the context to cancel the work. There still looks like a race condition on retErr as the above return on context cancel doesn't guarantee the write happens before the read.

File: private/buf/buflsp/mutex.go
Comment (emcfarlane) [inline private/buf/buflsp/mutex.go:81]: Maybe a `sync.Cond` could be used with `Broadcast` to wake all non owners and reschedule the lock and allow checking who owns the current lock:

```go
func (m *mutex) Lock(id uint64) {
    m.lock.Lock()
    for m.owner != 0 {
        if m.owner == id {
            m.lock.Unlock()
            panic("this id locked twice")
        }
        m.cond.Wait()
    }
    m.owner = id
    m.lock.Unlock()
}
func (m *mutex) Unlock(id uint64) {
    m.lock.Lock()
    if m.owner != id {
        m.lock.Unlock()
        panic("invalid")
    }
    m.id = 0
    m.cond.Broadcast()
    m.lock.Unlock()
}
```

File: private/buf/buflsp/mutex.go
Comment (emcfarlane) [inline private/buf/buflsp/mutex.go:155]: `mu.who.Store(0)` must be called with `mu.lock.Lock()` else you can race the `Lock()` function.

**Review by bufdev — COMMENTED** [review]
Oh my phone but don't forget changelog if we haven't done it

File: private/buf/buflsp/mutex.go
Comment (emcfarlane) [reply private/buf/buflsp/mutex.go:155]: Ah, this is guaranteed.

File: private/buf/buflsp/buflsp.go
Comment (mcy) [reply private/buf/buflsp/buflsp.go:207]: It waits for completion or deadline expiration. Your observation re: relying on the context to cancel the work is only correct if the work is I/O bound, but that is not the case here.

There is no data race; I suggest looking at how context cancelation and synchronization works (writing to a channel is synchronization).

File: private/buf/buflsp/mutex.go
Comment (mcy) [reply private/buf/buflsp/mutex.go:107]: Look for `unlock()` elsewhere in the PR. It is necessary to avoid squirrely control flow.

File: private/buf/buflsp/mutex.go
Comment (mcy) [reply private/buf/buflsp/mutex.go:81]: That's not what this is doing. This ensures that if A has locked X, it cannot also lock Y, because an acquire within a critical section is what causes deadlocks in the LSP.

File: private/buf/buflsp/mutex.go
Comment (mcy) [reply private/buf/buflsp/mutex.go:124]: Yeah that's just a missing branch.

File: private/buf/buflsp/mutex.go
Comment (emcfarlane) [reply private/buf/buflsp/mutex.go:81]: Why is the mutex not shared between X and Y?

File: private/buf/buflsp/buflsp.go
Comment (emcfarlane) [reply private/buf/buflsp/buflsp.go:207]: It's an unlikely data race, but the context can be canceled by the caller without `context.DeadlineExceeded ` so  it's still racy. CPU bounded can still take ctx and check to close work. Guess for a short lived CPU only this is fine, but most functions should take ctx and check it so it doesn't have to be only I/O bound work.

Comment (emcfarlane) [reply private/buf/buflsp/buflsp.go:207]: I think this could be written with an explicit channel.
```go
errChan := make(chan err, 1)
go func() { errChan <- work(ctx) }()
select {
case err <- errChan:
    return err
case <- ctx.Done():
    return ctx.Err()
}
```

File: private/buf/buflsp/mutex.go
Comment (emcfarlane) [inline private/buf/buflsp/mutex.go:120]: This now deadlocks on any request waiting for a lock as `pool.Lock()` is taken with `lock.Lock()` meaning all locks are blocked by the single lock waiting as the next locker waits on `pool.Lock()` which is blocked by `lock.Lock()`.

**Review by doriable — APPROVED** [review]

File: private/buf/buflsp/mutex.go
Comment (emcfarlane) [inline private/buf/buflsp/mutex.go:173]: Not really. If you are unable to swap it's confirmed to be poisoned as the lock is always held.
```go
if mu.who.CompareAndSwap(0, id) {
	panic("buflsp/mutex.go: non-reentrant lock locked twice by the same request")
}
```

File: private/buf/buflsp/mutex.go
Comment (emcfarlane) [inline private/buf/buflsp/mutex.go:168]: The `mu.lock.Unlock()` should be deferred, as this never gets called on `panic` of `storeWho`. So go routines will still deadlock on line `150` as they wait for this Unlock. Although does this matter? Would only matter if using recovers.

Comment (bufdev) [reply private/buf/buflsp/mutex.go:168]: @mcy can you check over all of your concurrency logic here before we merge?

File: private/buf/buflsp/file_manager.go
Comment (bufdev) [inline private/buf/buflsp/file_manager.go:30]: @doriable can you do a pass on naming here? None of these names match our style

File: private/buf/buflsp/buflsp.go
Comment (doriable) [inline private/buf/buflsp/buflsp.go:213]: Even though this is pretty small, I think it's reasonable to split this out in a different file, since this is fairly isolated functionality (even though I see the reasoning to keep it side-by-side with `adaptReplier`/`wrapReplier`).

File: private/buf/buflsp/buflsp.go
Comment (doriable) [inline private/buf/buflsp/buflsp.go:190]: I think the preferred naming convention in the code base for wrapping/wrapped types tend to be `wrapReplier` and `wrappedConnection` for the type below.

File: private/buf/buflsp/file.go
Comment (doriable) [inline private/buf/buflsp/file.go:81]: We might want to document what `version` here refers to and how it should be used.
And maybe `%s/text/content`, but I'm low confidence on this one (it makes sense to me either way, since this is the `string` representation of content).

File: private/buf/buflsp/file_manager.go
Comment (doriable) [inline private/buf/buflsp/file_manager.go:31]: `lsp` seems reasonable to me, but I think `table` should be more explicit, e.g. `fileURIToFile`, and there should be a comment indicating why we are using the `refcount` map rather than built-in map (I know there's a comment already on the `refcount` package about why to use it, but that isn't immediately reflected/translated here).
Similarly, we should be more explicit with `mutexPool` and call it `mutexPool` and add a comment to indicate this is a pool of the mutexes of the files that are managed by the `fileManager` and why it exists (to check for processes attempting to get a lock multiple times).

File: private/buf/buflsp/mutex.go
Comment (doriable) [inline private/buf/buflsp/mutex.go:57]: These are both helpers, so they can be at the bottom of the file, even though `withRequestID` is shared by another file. Also, the docs should be `%s/withReentrancy/withRequestID`.

File: private/buf/buflsp/report.go
Comment (doriable) [inline private/buf/buflsp/report.go:30]: Since this implements `reporter.Reporter`, it should probably be named `reporter`.

File: private/buf/buflsp/report.go
Comment (doriable) [inline private/buf/buflsp/report.go:64]: Doc update `%s/error2diagnostic/newDiagnostic`.

**Review by doriable — COMMENTED** [review]
Add some comments around style/readability.

File: private/buf/buflsp/report.go
Comment (doriable) [reply private/buf/buflsp/report.go:30]: Instead of alias-ing the package, let's leave this as `report` -- it is still a clear name for the implementation, since it is a "report" of the diagnostics, missing syntax, and unused imports.

We should probably rename `syntaxMissing` and `unusedImports` with the same map naming conventions, so in this case `fileNameToSyntaxMissing` and `fileNameToUnusedImports`.

Comment (doriable) [reply private/buf/buflsp/report.go:30]: Chatted in person, `syntaxMissing` is a good name for the set, and we'll rename `unusedImports`.

**Pull request merged**
